### ISSUE TITLE
Parsley is extremely, extremely brittle

### ISSUE THREAD

**Issue by ctrlcctrlv** [simoncozens/fontFeatures#26]
Everything about this code is a joy to work on except Parsley.

Parsley is extremely brittle. Just changing the order of statements fixes things and breaks others in cases when order obviously shouldn't matter.

Is it even possible to move to something else at this point?

**Comment by simoncozens** [comment]
I would love to and am open to suggestions. I haven't found anything else that's suitable, though. :-/

**Comment by ctrlcctrlv** [comment]
[Lark](https://github.com/lark-parser/lark)? I contributed to it a bit.

**Comment by ctrlcctrlv** [comment]
Parsley is so brittle that right now I've discovered a bug where in Pypy3 I get the dreaded Parsley non-error error `NameError: name 'metric' is not defined` while in CPython3 I get expected output. :facepalm:

**Comment by simoncozens** [comment]
Oh, interesting. So the main prerequisite is that the grammars need to be extensible *while* the parsing is happening. i.e. something which would not have parsed before now magically parses because we saw and executed a `LoadPlugin` statement. If lark can do that, let's go.

**Comment by simoncozens** [comment]
Also parsley's error/context handling *sucks* so that's another reason I want rid of it.

**Comment by ctrlcctrlv** [comment]
I can't even figure out how to see the grammars Parsley generates. It hides them from you and says they exist in the root directory which isn't true even at runtime. (I used GDB to be sure.)

**Comment by ctrlcctrlv** [comment]
As far as `LoadPlugin` goes, I might have to try to write a Lark minimal example to see, I don't know offhand if it does, but why not just do some look ahead for `LoadPlugin`'s?

**Comment by ctrlcctrlv** [comment]
I can't think of a way to do that in Lark (doesn't mean it's not possible, @erezsh would know), but I think it's better anyway to just run through the file twice. First look for all the `LoadPlugin`'s and load them, then parse the file. Hell, I'd even say, for sanity, don't allow `LoadPlugin` anywhere but at the top. Plenty of languages do this. There's no really good reason except "wow magic" to do it any other way.

**Comment by ctrlcctrlv** [comment]
If we're trading this minor feature (ooo we can load plugins from the middle of the file fancy) for the hellworld that is Parsley I vote for Lark please :joy:

**Comment by simoncozens** [comment]
Yeah, that's fine. We can even regex search for `LoadPlugin`s, whatever. Looks like it would be a lot of work, but worth doing.

**Comment by ctrlcctrlv** [comment]
Here is a Lark base grammar for FEE.

    ?start: statement+

    statement: (load_plugin_statement | regex_statement) ";"

    load_plugin_statement: LOAD_PLUGIN_VERB IDENTIFIER
    LOAD_PLUGIN_VERB: "LoadPlugin"
    IDENTIFIER: /[A-Za-z0-9-_.]/+

    regex_statement: REGEX_VERB glyphselector+
    REGEX_VERB: "Regex"

    STARTGLYPHNAME: /[abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_]/
    MIDGLYPHNAME: STARTGLYPHNAME | /[.-]/
    BARENAME: STARTGLYPHNAME MIDGLYPHNAME*
    CLASS: CLASSNAME | INLINE_CLASS
    INLINE_CLASS: "[" (WS* (CLASSNAME | BARENAME))* "]"
    CLASSNAME: "@" BARENAME

    ANYTHING: /[^\s]/
    REGEX: "/" ANYTHING* "/"

    HEXDIGIT: /[0-9A-Fa-f]/
    UNICODECHAR: "U+" HEXDIGIT~1..6
    UNICODERANGE: UNICODECHAR "=>" UNICODECHAR

    GLYPHSUFFIX: ("." | "~") STARTGLYPHNAME+
    glyphselector: (UNICODERANGE | UNICODECHAR | REGEX | BARENAME | CLASS) GLYPHSUFFIX*

    valuerecord: SIGNED_NUMBER | FEE_VALUE_RECORD | FEA_VALUE_RECORD
    FEE_VALUE_RECORD: "<" ( FEE_VALUE_VERB "=" SIGNED_NUMBER )+ ">"
    FEE_VALUE_VERB: "xAdvance" | "xPlacement" | "yAdvance" | "yPlacement"
    FEA_VALUE_RECORD: "<" SIGNED_NUMBER SIGNED_NUMBER SIGNED_NUMBER SIGNED_NUMBER ">"

    %import python(COMMENT)
    %import common(ESCAPED_STRING, SIGNED_NUMBER, LETTER, WS)
    %ignore WS
    %ignore COMMENT

**Comment by ctrlcctrlv** [comment]
So basically what needs to be done is this needs to run through several `lark.Transformer`'s (or similar):

1. one to validate the inputs
2. one to transform the input into the correct Python object

Parsley gave us those for free, but it's not so hard for me to implement.

**Comment by ctrlcctrlv** [comment]
Oh, actually there should be a "zeroth" transformer, to run through the token tree and look for `load_plugin_statement`'s, load them, rebuild the grammar, and rebuild the tree. Yep.

**Comment by ctrlcctrlv** [comment]
@simoncozens What I'm working on right now side steps the whole issue. You can `LoadPlugin` anywhere and it'll work, because I am treating it as a bunch of mini-parsers, not one big parser. Each plugin gets its own parser, and Verb's decide which one to use.

```python3
import feeLib

sample_file = """
#comment ca va
#   a
LoadPlugin Variables;
LoadPlugin ClassDefinition;
DefineClass @low = lower | higher;
DefineClass @QQ = @XX;
DefineClass @derp = /.derp$/;
DefineClass @derp = U+30=>U+39;
DefineClass @wide = width > 900;
DefineClass @em = width = 1000;

Set once = 11;

IDontExist @get @owned.hard [l o l] [@yes @no] [] [@QQ] U+20 @cluck~cluck.cluck;

Routine {
    Substitute a => b;
    Substitute b => c;

    Routine {
        Substitute a => b;
    };
};

"""

fp = feeLib.FeeParser()
ps = fp.parseString(sample_file)
print(ps)
```

```python3
# Unknown verb: IDontExist
# Unknown verb: Substitute
# Unknown verb: Routine
[('LoadPlugin', 'Variables'), ('LoadPlugin', 'ClassDefinition'), ('DefineClass', ('@low', [Tree('primary', [Tree('conjunction', [Tree('primary', [GlyphSelector<lower>]), Token('CONJUNCTOR', '|'), Tree('primary', [GlyphSelector<higher>])])])])), ('DefineClass', ('@QQ', [Tree('primary', [GlyphSelector<@XX>])])), ('DefineClass', ('@derp', [Tree('primary', [GlyphSelector</.derp$/>])])), ('DefineClass', ('@derp', [Tree('primary', [GlyphSelector<U+0030=>U+003A>])])), ('DefineClass', ('@wide', [Tree('primary', [Tree('predicate', [Tree('comparison', [Token('METRIC', 'width'), Token('COMPARATOR', '>'), Tree('comp_value', [Token('SIGNED_NUMBER', '900')])])])])])), ('DefineClass', ('@em', [Tree('primary', [Tree('predicate', [Tree('comparison', [Token('METRIC', 'width'), Token('COMPARATOR', '='), Tree('comp_value', [Token('SIGNED_NUMBER', '1000')])])])])])), ('Set', ('once', '11')), ('IDontExist', ['@get', '@owned.hard', '[l', 'o', 'l]', '[@yes', '@no]', '[]', '[@QQ]', 'U+20', '@cluck~cluck.cluck']), ('Routine', [('Substitute', ['a', '=>', 'b']), ('Substitute', ['b', '=>', 'c']), ('Routine', [('Substitute', ['a', '=>', 'b'])])])]
```

Obviously more work to be done here. `Variables` etc will become default loaded of course, just for now they're not since I'm testing.

**Comment by ctrlcctrlv** [comment]
This is the first FEA output from my new Lark implementation.

`DefineClass` is 100% implemented!

```fea
LoadPlugin ClassDefinition;
DefineClass @derp = /^f/;
DefineClass @low = a | c;
DefineClass @low2 = a | c | e;
DefineClass @QQ = @low2 | @derp;
DefineClass @derp = U+30=>U+39;
DefineClass @pred = f & hasglyph(/f/ g);
DefineClass @wide = width > 900;
DefineClass @wide_caps = /^[A-Z]$/ & width > 450;
DefineClass @em = width = 1000;
DefineClass @inline = [l o l]; # works because HelperTransformer in use
```

```fea
@derp = [zero one two three four five six seven eight nine];
@low = [c a];
@low2 = [c a e];
@QQ = [finalmem f c four finalpedagesh e finalpe fiveeighths a finaltsadi finalnun fraction five finalkafdagesh finalmemwide finalkaf];
@pred = [f];
@wide = [percent threequarters uni01C4 uni01C5 uni01C6 uni01CA uni01F1 uni01F2 uni01F3 uni0409 uni0416 uni04C1 uni04DC uni2031 uni2156 uni2157 uni2158 uni215A threeeighths fiveeighths seveneighths uni2166 uni2167 uni216B uni2176 uni2177 uni217B uniFB03 uniFB04];
@wide_caps = [A B C D G H J K M N O P Q R S T U V W X Y Z];
@em = [];
@inline = [l o l];
```

So happy!!!

**Comment by ctrlcctrlv** [comment]
Also I have to say @simoncozens, you may not believe me, but I never once received a cryptic error from Lark while implementing this. Not once. All errors were clear, all mistakes obviously mine. I love it so much more than Parsley.
